(Publish date: on hold)  - Cumulative metrics transition guide (#11876)
newrelic/docs-website: src/content/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/cumulative-metrics-transition-guide.mdx

---
title: Transition guide for OpenTelemetry cumulative sum metrics
metaDescription: 'A transition guide about a change in how New Relic handles OpenTelemetry cumulative sum metrics, and how to update your queries and alerts for that change.'
redirects: 
---

Our OTLP metric pipeline is being upgraded to better support how we handle [cumulative metrics](/docs/data-apis/understand-data/metric-data/cumulative-metrics). As part of this upgrade we're transitioning the underlying [metric data type](/docs/data-apis/understand-data/metric-data/metric-data-type/#metric-types) for [monotonic cumulative sums](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums) from a `gauge` type to the `cumulativeCount` type. The `cumulativeCount` type allows you to query both the cumulative and delta value associated with reported data. 

> Starting March 14, 2023, new New Relic customers will use the new `cumulativeCount` type, but if you were a New Relic customer before that date, you won't have access to that metric type until June 13. On that date, the use of our `gauge` metric type for cumulative sums will be deprecated and reach end-of-life (EoL), and your cumulative sums will be stored as the `cumulativeCount` type. 

Reviewer: I stumbled a bit over the "new New Relic" phrase. How about revising like this: Starting March 14, 2023, customers who are getting started with New Relic...."

Also, will New Relic be converting my existing monotonic cumulative sum data in NRDB, so that if I want to query historical sums, I'll need to modify my queries for cumulativeCount? Or, is this just for all future sums?

Reviewer 2: Existing monotonic cumulative sum data that was ingested prior to the EoL (i.e. stored as a `gauge` type in NRDB) will not be converted after the fact. Those data points will be retained per the defined retention duration in the account, so they will continue to be available for that period. To view both datatypes in a single NRQL view, something like the following could be used:
**cumulativeCount Query 1**
```
FROM Metric SELECT rate(sum(<metricName>), 1 minute) as 'delta rate', latest(<metricName>['cumulative']) as 'cumulative' WHERE <metricName>[type] = 'cumulativeCount'  TIMESERIES 
```
**gauge Query 2**
```
FROM Metric SELECT derivative(<metricName>, 1 minute) as 'gauge derivative', latest(<metricName>) as 'latest' WHERE <metricName>[type] = 'gauge' TIMESERIES 
```

I've tested these queries across the metric type boundary and they return as you would expect. The examples are bit verbose so I was hesitated to include them anywhere.

If you were a customer using cumulative metrics before March 14, you should check and potentially modify existing queries and alerts that may be relying on those metrics. The sections below will teach you: 

* How to know if your account is impacted
* How to prepare for the EoL
* How to preview the `cumulativeCount` functionality 

<Callout variant="important">
Non-monotonic cumulative sums will **not** be affected by this upgrade.
</Callout>

## Check your account for impact [#impact]

> If your account was sending OTLP cumulative sum metrics to New Relic before March 14, 2023, then that account will be impacted by the EoL on June 13, 2023. 

Reviewer: Just to be safe, I'd add "monotonic": If your account was sending OTLP monotonic cumulative sum...."

Author: @tclayton-newr wanted to get your take on this. should we always be mentioning 'monotonic' when we reference this metric?

Reviewer 2: Yes, while more verbose that is the most accurate statement of metrics that are affected by this EoL

To check if your account is impacted, run the following NRQL query:

```
FROM NrIntegrationError SELECT count(*) 
WHERE newRelicFeature = 'cumulativeSumAsGaugeEoL' 
SINCE 24 hours ago
```

The query will return results if you've reported metrics impacted by this EoL in the last 24 hours. 

To investigate further, you can view the event message to see which metrics are affected:

```
FROM NrIntegrationError SELECT metricName 
WHERE newRelicFeature = 'cumulativeSumAsGaugeEoL' 
LIMIT MAX SINCE 24 hours ago
```

Any monitoring or alerting involving these metrics will need to be reviewed and updated prior to the EoL date. The next section contains suggested actions for impacted metrics. 

> ## What to do if you're impacted [#what-to-do]

Reviewer: It sounds like older users won't have access to this new metric type until midnight on June 12th. If you were a user, how would you time this rollout? Do you have to stage these query changes and then roll them out at midnight on June 12th? 

I see later that the doc says older users can access this new metric type by selecting the default delta value. So, are we saying that users could convert all their queries ahead of time and they will work? Are we saying the only difference is how we ingest new sums?

OK, later I see that we're telling users to create side-by-side queries for alerts so that when the switch is flipped, there won't be any impact. Could you make that much clearer earlier?

> We recommend impacted accounts to check and potentially modify existing NRQL-generated charts and alerts that target OTLP cumulative sum metrics. Our [OpenTelemetry best practices](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/best-practices/opentelemetry-best-practices-metrics#query) outline the differences in querying between a `gauge` and a `cumulativeCount` metric type. Namely, you'll be able to target either the delta value (by default) or the cumulative value by selecting the cumulative field. 

Reviewer: How about a revision like this: If you have any affected accounts, we recommend you check, and potentially modify, existing NRQL-generated charts and alerts that target OTLP cumulative sum metrics.

Below we focus on updating [alert conditions](/docs/alerts-applied-intelligence/new-relic-alerts/alert-conditions/create-nrql-alert-conditions), because alerts are generally more important than charts. For your custom charts, you can either:

* Take the same general approach we describe for alerts: create a new chart with an updated query and erase the old chart after the EoL, OR
* Wait until after the EoL and replace the queries

## Suggested action: Create new versions of alerts [#alerts]

> Here are some examples of queries that might be used in alerts. We'll give you the original query, then a `gauge` version of the query, and a `cumulativeCount` version of the query. You should replace your existing alert with the `gauge` version of the query, and create a duplicate alert condition that uses the `cumulativeCount` version. That way, when we transition your account to `cumulativeCount` on June 16, that alert condition will be available when the old query no longer works. 

Reviewer: Here are some questions:

* How about calling "original query" something like the generic query or starting query?
* Why does this say that you should replace your existing alert with the `gauge` version of the query? Weren't existing users already using gauges in their queries since monotonic cumulative sums used to be converted to gauges? I was just expecting to be told to add a duplicate alert condition that uses cumulativeCount.

Author: @tclayton-newr question for you: why do we tell them to replace the existing one with the `gauge`? I don't really get that either; wouldn't it work remaining as is? Why does the existing query need to change?

Reviewer 2: That's a great question! The existing query needs to target the `gauge` type specifically because if the account updates to use the `cumulativeCount` type, the default value returned by the query will change. For example, consider a metric reporting uptime (in seconds) of a server every second (a counter increasing by 60 every minute) and we'll assume we're at 120 seconds of uptime. 
Then the `gauge` query 
```
FROM Metric SELECT latest(<metricName>)
```
would return 120 since (prior to this EoL) that was the best value we could return; the latest cumulative value the account sent us. 

However, the `cumulativeCount` version of the same query would now be targeting a **delta** value instead of the cumulative value. So it would return a value of 1, because the delta rate of change is likely a value of 1 every second. While this example is somewhat contrived (a customer likely isn't alerting based on uptime of their servers) it does show how a query can change from reporting an ever-increasing counter to just the most recent delta value, which can be two very different results.

WHERE <metricName>[type] = 'gauge'
```

Cumulative count version of query:

```
FROM Metric SELECT rate(sum(<metricName>[cumulative]), 1 minute) 
WHERE <metricName>[type] = 'cumulativeCount'
```

## Suggested action: preview cumulativeCount before the EoL [#preview]

Affected accounts will continue to have their cumulative sums stored as `gauge` types until the EoL. However, optionally, you can try out `cumulativeCount` before the EoL to validate your queries and alerts are behaving as expected. To test `cumuativeCount` types early, add the following attribute to OTLP data you send to New Relic:

```
newrelic_metric_type=cumulativeCount
```

<Callout variant="important">
When you start sending `newrelic_metric_type=cumulativeCount` on your data, any cumulative sums we receive will be translated into `cumulativeCount` types. This means any existing queries or alerts set up to target a `gauge` type will no longer be valid. Keep this in mind when testing the `cumulativeCount` experience. To return to the `gauge` type, simply stop setting the `newrelic_metric_type` attribute. 
</Callout>

### Instrumentation library [#instrumentation-library]

> If you're using an OTLP instrumentation library within your code base, you should reference the [appropriate language SDK](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/#instrument) for your application. Any method of adding the attribute `newrelic_metric_type=cumulativeCount` to individual metric data points or to the `Resource attribute` level should be sufficient. 

Reviewer 2: ```suggestion
If you're using an OTLP instrumentation library within your code base, you should reference the [appropriate language SDK](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/#instrument) for your application. Any method of adding the attribute `newrelic_metric_type=cumulativeCount` to individual metric data points or to the resource attribute level should be sufficient. 
```

I think referring to this in plain text makes more sense. In the draft I think I used monospace so that is my fault 😅

---
title: Transition guide for OpenTelemetry cumulative sum metrics
metaDescription: 'A transition guide about a change in how New Relic handles OpenTelemetry cumulative sum metrics, and how to update your queries and alerts for that change.'
redirects: 
---

Our OTLP metric pipeline is being upgraded to better support how we handle [cumulative metrics](/docs/data-apis/understand-data/metric-data/cumulative-metrics). As part of this upgrade we're transitioning the underlying [metric data type](/docs/data-apis/understand-data/metric-data/metric-data-type/#metric-types) for [monotonic cumulative sums](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums) from a `gauge` type to the `cumulativeCount` type. The `cumulativeCount` type allows you to query both the cumulative and delta value associated with reported data. 

> Starting March 14, 2023, new accounts will have access to this new `cumulativeCount` type, but if your account existed before that date, you won't have access to that metric type until June 13. On that date, the use of our `gauge` metric type for cumulative sums will be deprecated and reach end-of-life (EoL), and your cumulative sums will be stored as the `cumulativeCount` type. 

Reviewer 2: ```suggestion
Starting March 14, 2023, new accounts will have access to this new `cumulativeCount` type, but if your account was sending cumulative monotonic sums prior to March 14th, you won't have access to that metric type until June 13. On that date, the use of our `gauge` metric type for cumulative sums will be deprecated and reach end-of-life (EoL), and your cumulative sums will be stored as the `cumulativeCount` type. 
```

**This 90 day EoL period is to give you time to update queries and alerts that make use of the affected metrics.** The sections below will teach you: 

* How to know if your account is impacted
* How to prepare for the EoL
* How to preview the `cumulativeCount` functionality 

<Callout variant="important">
Non-monotonic cumulative sums will **not** be affected by this upgrade.
</Callout>

## Check your account for impact [#impact]

If your account was sending OTLP cumulative sum metrics to New Relic before March 14, 2023, then that account will be impacted by the EoL on June 13, 2023. 

To check if your account is impacted, run the following NRQL query:

```
FROM NrIntegrationError SELECT count(*) 
WHERE newRelicFeature = 'cumulativeSumAsGaugeEoL' 
SINCE 24 hours ago
```

The query will return results if you've reported metrics impacted by this EoL in the last 24 hours. 

To investigate further, you can view the event message to see which metrics are affected:

```

> FROM NrIntegrationError SELECT metricName 

Reviewer 2: ```suggestion
FROM NrIntegrationError SELECT uniques(metricName)
```
My fault since I provided the original query 😛

```

Gauge version of the query:
``` 
FROM Metric SELECT latest(<metricName>) 
WHERE <metricName>[type] = 'gauge'
```

Cumulative count version of the query: 
```
FROM Metric SELECT latest(<metricName>[cumulative]) 
WHERE <metricName>[type] = 'cumulativeCount'
```

### Rate of change queries [#rate-of-change-queries]

Original query: 

```
FROM Metric SELECT derivative(<metricName>, 1 minute)
```

Gauge version of query:
``` 
FROM Metric SELECT derivative(<metricName>, 1 minute) 
WHERE <metricName>[type] = 'gauge'
```

Cumulative count version of query:

```
FROM Metric SELECT rate(sum(<metricName>[cumulative]), 1 minute) 
WHERE <metricName>[type] = 'cumulativeCount'
```

## Suggested action: preview cumulativeCount before the EoL [#preview]

Affected accounts will continue to have their cumulative sums stored as `gauge` types until the EoL. However, optionally, you can try out `cumulativeCount` before the EoL to validate your queries and alerts are behaving as expected. To test `cumuativeCount` types early, add the following attribute to OTLP data you send to New Relic:

```
newrelic_metric_type=cumulativeCount
```

<Callout variant="important">

> When you start sending `newrelic_metric_type=cumulativeCount` on your data, any cumulative sums we receive will be translated into `cumulativeCount` types. This means any existing queries or alerts set up to target a `gauge` type will no longer be valid. Keep this in mind when testing the `cumulativeCount` experience. To return to the `gauge` type, simply stop setting the `newrelic_metric_type` attribute. 

Reviewer 2: ```suggestion
When you start sending `newrelic_metric_type=cumulativeCount` on your data, any cumulative monotonic sums we receive will be translated into `cumulativeCount` types. This means any existing queries or alerts set up to target a `gauge` type will no longer be valid. Keep this in mind when testing the `cumulativeCount` experience. To return to the `gauge` type, simply stop setting the `newrelic_metric_type` attribute. 
```

---
title: Transition guide for OpenTelemetry cumulative sum metrics
metaDescription: 'A transition guide about a change in how New Relic handles OpenTelemetry cumulative sum metrics, and how to update your queries and alerts for that change.'
redirects: 
---

Our OTLP metric pipeline is being upgraded to better support how we handle [cumulative metrics](/docs/data-apis/understand-data/metric-data/cumulative-metrics). As part of this upgrade we're transitioning the underlying [metric data type](/docs/data-apis/understand-data/metric-data/metric-data-type/#metric-types) for [monotonic cumulative sums](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums) from a `gauge` type to the `cumulativeCount` type. The `cumulativeCount` type allows you to query both the cumulative and delta value associated with reported data. 

> Starting March 14, 2023, new accounts will have access to this new `cumulativeCount` type, but if your account was sending cumulative monotonic sums prior to March 14th, you won't have access to that metric type until June 13. On that date, the use of our `gauge` metric type for cumulative sums will be deprecated and reach end-of-life (EoL), and your cumulative sums will be stored as the `cumulativeCount` type. 

Reviewer 2: ```suggestion
Starting March 14, 2023, we will ingest cumulative monotonic metrics according to the following end-of-life (EoL) timeline:
- March 14th, 2023: The ingest of cumulative monotonic sums as the `cumulativeCount` metric type is released. The prior experience of ingesting these metrics as the `gauge` metric type is deprecated.
  - Accounts that were actively sending cumulative monotonic sums prior to this date will continue to use `gauge` type ingest. 
  - All other accounts will have cumulative monotonic sums ingested as `cumulativeCount` metric types. 
  
- March 14th - June 13th, 2023: Accounts using the deprecated `gauge` type ingest may opt-in to `cumulativeCount` type ingest at any time.

- June 13th, 2023: All accounts will be migrated to `cumulativeCount` type ingest for cumulative monotonic sums.
```
The formatting here is letting me down a bit but does this make sense? The idea was to make the dates/events a bit more readable.

**This 90 day EoL period is to give you time to update queries and alerts that make use of the affected metrics.** The sections below will teach you: 

* How to know if your account is impacted
* How to prepare for the EoL

> * How to preview the `cumulativeCount` functionality 

Reviewer 2: ```suggestion
* How to opt-in to the `cumulativeCount` functionality prior to the EoL date
```
Goal is to make this sound less like an early preview and more like switching over to the new experience sooner.

> <Callout variant="important">
> Cumulative sums that are non-monotonic will **not** be affected.
> </Callout>

Reviewer 2: ```suggestion
<Callout variant="important">
Early opt-in to `cumulativeCount` type ingest is the best way to ensure affected queries and alerts are not disrupted by the EoL.
</Callout>

<Callout variant="info">
Cumulative sums that are non-monotonic will **not** be affected.
</Callout>
```

Goal is to try and endorse the early opt-in as early in the doc as possible.

Reviewer 2: Also I have no idea if `variant="info"` is a valid value at all 😅

## Check your account for impact [#impact]

If your New Relic account was sending OTLP cumulative sum metrics to New Relic before March 14, 2023, then you'll be impacted by the EoL on June 13, 2023. 

To check if your account is impacted, run the following NRQL query:

```
FROM NrIntegrationError SELECT count(*) 
WHERE newRelicFeature = 'cumulativeSumAsGaugeEoL' 
SINCE 24 hours ago
```

The query will return results if you've reported metrics impacted by this EoL in the last 24 hours. 

To investigate further, you can view the event message to see which metrics are affected:

```
FROM NrIntegrationError SELECT uniques(metricName)
WHERE newRelicFeature = 'cumulativeSumAsGaugeEoL' 
LIMIT MAX SINCE 24 hours ago
```

Any monitoring or alerting involving these metrics will need to be reviewed and updated prior to the EoL date. The next section contains suggested actions for impacted metrics. 

## What to do if you're impacted [#what-to-do]

> On June 13, the reporting of OTLP monotonic cumulative sum metrics will switch from using the `gauge` to `cumulativeCount`. If you're using those metrics, you'll want to prepare for that. For example, you may want to create duplicate alert conditions with the new metric type so that, when the switch happens, you won't have a disruption (and we explain how to do that below).

Reviewer 2: ```suggestion
On June 13, the reporting of OTLP monotonic cumulative sum metrics will switch from using the `gauge` to `cumulativeCount` metric type. If you're using those metrics, you'll want to prepare for that. For example, you may want to create duplicate alert conditions with the new metric type so that, when the switch happens, you won't have a disruption (and we explain how to do that below).
```

Our [OpenTelemetry best practices](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/best-practices/opentelemetry-best-practices-metrics#query) outline the differences in querying between a `gauge` and a `cumulativeCount` metric type. Namely, you'll be able to target either the delta value (by default) or the cumulative value by selecting the cumulative field.

> Let's take an example query. Up until the EoL, a query like: 

Reviewer 2: ```suggestion
Let's take an example query of a cumulative monotonic sum metric `metricName`. Up until the EoL, a query like: 
```

```
FROM Metric SELECT latest(<metricName>)
```

> would be evaluating only `gauge` metrics. Once the EoL happens, it would be evaluating both `gauge` and `cumulativeCount` metrics. Depending on the alert, that could cause false results.

Reviewer 2: ```suggestion
would be evaluating only `gauge` metrics. Once the EoL happens, the underlying metric type associated with `metricName` will change to `cumulativeCount`. Depending on the alert, this change could cause false results as the metric value for a `gauge` and `cumulativeCount` type represent different values of the data point. 
```

```

Gauge version of the query:
``` 
FROM Metric SELECT latest(<metricName>) 
WHERE <metricName>[type] = 'gauge'
```

Cumulative count version of the query: 
```
FROM Metric SELECT latest(<metricName>[cumulative]) 
WHERE <metricName>[type] = 'cumulativeCount'
```

### Rate of change queries [#rate-of-change-queries]

Original query: 

```
FROM Metric SELECT derivative(<metricName>, 1 minute)
```

Gauge version of query:
``` 
FROM Metric SELECT derivative(<metricName>, 1 minute) 
WHERE <metricName>[type] = 'gauge'
```

Cumulative count version of query:

```
FROM Metric SELECT rate(sum(<metricName>[cumulative]), 1 minute) 
WHERE <metricName>[type] = 'cumulativeCount'
```

## Suggested action: preview cumulativeCount before the EoL [#preview]

Affected accounts will continue to have their cumulative sums stored as `gauge` types until the EoL. However, optionally, you can try out `cumulativeCount` before the EoL to validate your queries and alerts are behaving as expected. To test `cumuativeCount` types early, add the following attribute to OTLP data you send to New Relic:

```
newrelic_metric_type=cumulativeCount
```

<Callout variant="important">

> When you start sending `newrelic_metric_type=cumulativeCount` on your data, any cumulative monotonic sums we receive will be translated into `cumulativeCount` types. This means any existing queries or alerts set up to target a `gauge` type will no longer be valid. Keep this in mind when testing the `cumulativeCount` experience. To return to the `gauge` type, simply stop setting the `newrelic_metric_type` attribute. 

Reviewer 2: ```suggestion
When you start sending `newrelic_metric_type=cumulativeCount` on your data, any cumulative monotonic sums we receive will be translated into `cumulativeCount` types. This means any existing queries or alerts set up to target a `gauge` type will receive unexpected results. Keep this in mind when testing the `cumulativeCount` experience. To return to the `gauge` type, simply stop setting the `newrelic_metric_type` attribute. 
```
Slightly more correct.

Our OTLP metric pipeline is being upgraded to better support how we handle [cumulative metrics](/docs/data-apis/understand-data/metric-data/cumulative-metrics). As part of this upgrade we're transitioning the underlying [metric data type](/docs/data-apis/understand-data/metric-data/metric-data-type/#metric-types) for [monotonic cumulative sums](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums) from a `gauge` type to the `cumulativeCount` type. The `cumulativeCount` type allows you to query both the cumulative and delta value associated with reported data. 

Starting March 14, 2023, we will ingest cumulative monotonic metrics according to the following end-of-life (EoL) timeline:

* March 14, 2023: The ingest of cumulative monotonic sums as the `cumulativeCount` metric type is released (non-monotonic cumulative sums are **not** affected). The prior experience of ingesting these metrics as the `gauge` metric type is deprecated.
* Accounts that were actively sending cumulative monotonic sums prior to this date can continue to use `gauge` type ingest. 

> * All other accounts will have cumulative monotonic sums ingested as `cumulativeCount` metric types. 

Reviewer 2: Can these bullet points be nested under the "March 14th, 2023" parent?

* March 14 through June 13, 2023: Accounts using the deprecated `gauge` type ingest can opt-in to `cumulativeCount` type ingest. **Early opt-in is the best way to ensure affected queries and alerts are not disrupted by the EoL**.
* June 13, 2023: All accounts will be migrated to `cumulativeCount` type ingest for cumulative monotonic sums.

> If you were ingested cumulative monotonic sums prior to March 14, this 90 day period gives you time to update queries and alerts that make use of the affected metrics.

Reviewer 2: ```suggestion
If you were ingesting cumulative monotonic sums prior to March 14, this 90 day period gives you time to update queries and alerts that make use of the affected metrics.
```
